frigate/record/cleanup.py

```diff
@@ -33,10 +33,153 @@ def clean_tmp_clips(self) -> None:
+                Recordings.motion,
+            )
+            .where(
+                Recordings.camera == config.name,
+                Recordings.end_time < expire_date,
+            )
+            .order_by(Recordings.start_time)
+            .namedtuples()
+            .iterator()
+        )
+
+        # loop over recordings and see if they overlap with any non-expired events
+        # TODO: expire segments based on segment stats according to config
+        event_start = 0
+        deleted_recordings = set()
+        for recording in recordings:
+            keep = False
+            # Now look for a reason to keep this recording segment
+            for idx in range(event_start, len(events)):
+                event: Event = events[idx]
+
+                # if the event starts in the future, stop checking events
+                # and let this recording segment expire
+                if event.start_time > recording.end_time:
+                    keep = False
+                    break
+
+                # if the event is in progress or ends after the recording starts, keep it
+                # and stop looking at events
+                if event.end_time is None or event.end_time >= recording.start_time:
+                    keep = True
+                    break
+
+                # if the event ends before this recording segment starts, skip
+                # this event and check the next event for an overlap.
+                # since the events and recordings are sorted, we can skip events
+                # that end before the previous recording segment started on future segments
+                if event.end_time < recording.start_time:
+                    event_start = idx
+
+            # Delete recordings outside of the retention window or based on the retention mode
+            if (
+                not keep
+                or (
+                    config.record.events.retain.mode == RetainModeEnum.motion
+                    and recording.motion == 0
+                )
+                or (
+                    config.record.events.retain.mode == RetainModeEnum.active_objects
+                    and recording.objects == 0
+                )
+            ):
+                Path(recording.path).unlink(missing_ok=True)
+                deleted_recordings.add(recording.id)
+
+        logger.debug(f"Expiring {len(deleted_recordings)} recordings")
+        # delete up to 100,000 at a time
+        max_deletes = 100000
+        deleted_recordings_list = list(deleted_recordings)
+        for i in range(0, len(deleted_recordings_list), max_deletes):
+            Recordings.delete().where(
+                Recordings.id << deleted_recordings_list[i : i + max_deletes]
+            ).execute()
+
+    def expire_existing_camera_previews(
+        self, expire_date: float, config: CameraConfig, events: Event
+    ) -> None:
+        """Delete previews for existing camera based on record retention config."""
```

Comment:
If the plan is to retain previews in sync with recordings, its _probably_ simpler and faster to just check for recordings in the database that overlap with previews in the db. It would also be more efficient to merge this function into the recording expiration to avoid pulling all the data in to memory a second time.

Reply:
yeah good point, I think the way I came up with will be a lot more efficient and easier to understand